### PR TITLE
Userland: Add less style terminal pager

### PR DESCRIPTION
**Pull request by Petelliott**
I've implemented the less terminal pager for serenity, and made man use it by default.

### Features:
- Scrollback
- Custom prompts using less's prompt format language.
- More emulation when executed as more.
- Runs in an alternate screen by default.
- Handles ANSI escape sequences in line wrapping logic.

Since this isn't really less, but serenity own pager, i'm happy to rename it to something like 'page' or 'pager' if that is preferred.

2021-07-06-015126_1920x1080_scrot

### CODE REVIEW

File: Userland/Utilities/less.cpp [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+        m_line += write_range(min(m_line + m_height - 1, m_line + (m_lines.size() - m_line)), n);
+        status_line();
+
+        fflush(m_tty);
+    }
+
+    void init()
+    {
+        while (m_lines.size() < m_height) {
+            read_line();
```
Comment (alimpfard) [inline Userland/Utilities/less.cpp:157]: What if we don't have enough lines to fill the screen?
won't this loop forever?

File: Userland/Utilities/less.cpp [hunk 2]
```diff
@@ -173,7 +173,19 @@ class Pager {
 
     void status_line()
     {
-        out(m_tty, "\e[7m -- less -- (line {})\e[27m", m_line + 1);
+        out(m_tty, "\e[7m ");
+        render_status_line(m_prompt);
+        out(m_tty, " \e[27m");
+    }
+
+    void set_filename(const StringView& filename)
```
Comment (alimpfard) [inline Userland/Utilities/less.cpp:186]: Please use east-const as per the project coding style.

File: Userland/Utilities/less.cpp [hunk 3]
```diff
@@ -190,7 +202,52 @@ class Pager {
...
+                case 'f':
+                    off = render_status_line(prompt, off + 1, ':', m_file == stdin);
+                    off = render_status_line(prompt, off + 1, '.', m_file != stdin);
+                    break;
+                case 'e':
+                    off = render_status_line(prompt, off + 1, ':', !at_end());
+                    off = render_status_line(prompt, off + 1, '.', at_end());
+                    break;
+                default:
+                    VERIFY_NOT_REACHED();
```
Comment (alimpfard) [inline Userland/Utilities/less.cpp:234]: what if someone just puts a `?` at the end?
maybe let's not crash?

File: Userland/Utilities/less.cpp [hunk 4]
```diff
@@ -214,8 +271,11 @@ static String get_key_sequence()
 int main(int argc, char** argv)
 {
     const char* filename = "-";
+    const char* prompt = "?f%f :.(line %l)?e (END):.";
```
Comment (alimpfard) [inline Userland/Utilities/less.cpp:282]: It'd be nice to have a manpage for this, and it'd be even nicer if that manpage explains this format (or rather, what we support of it)

File: Userland/Utilities/less.cpp [hunk 5]
```diff
@@ -272,10 +278,19 @@ int main(int argc, char** argv)
 {
     const char* filename = "-";
     const char* prompt = "?f%f :.(line %l)?e (END):.";
+    bool dont_switch_buffer = false;
+    bool quit_at_eof = false;
+    bool emulate_more = false;
+
+    if (LexicalPath(argv[0]).basename() == "more")
```
Comment (alimpfard) [inline Userland/Utilities/less.cpp:287]:
```suggestion
    if (LexicalPath::basename(argv[0]) == "more"sv)
```

File: Userland/Utilities/less.cpp [hunk 6]
```diff
@@ -276,6 +276,8 @@ static String get_key_sequence()
 
 int main(int argc, char** argv)
 {
+    VERIFY(isatty(STDOUT_FILENO));
```
Comment (alimpfard) [inline Userland/Utilities/less.cpp:279]: Maybe just...don't do anything instead? (i.e. be `cat`)

File: Userland/Utilities/man.cpp [hunk 7]
```diff
@@ -78,12 +114,18 @@ int main(int argc, char* argv[])
     auto file = Core::File::construct();
     file->set_filename(make_path(section));
 
+    String pager_command = pager;
+    if (!pager) {
+        pager_command = String::formatted("less -P 'Manual Page {}({}) line %l?e (END):.'", name, section);
```
Comment (alimpfard) [inline Userland/Utilities/man.cpp:119]: That's not nice, if `name` contains a quote, this is a syntax error, or a borked shell injection path :P

File: Userland/Utilities/man.cpp [hunk 8]
```diff
@@ -99,4 +141,8 @@ int main(int argc, char* argv[])
 
     String rendered = document->render_for_terminal(view_width);
     out("{}", rendered);
+
+    fclose(stdout);
+    int wstatus;
+    waitpid(pager_pid, &wstatus, 0);
```
Comment (alimpfard) [inline Userland/Utilities/man.cpp:147]: what if someone `^Z`'s the pager?

**Review by alimpfard — COMMENTED** [review]
Looks pretty nice!

File: Userland/Utilities/less.cpp (hunk 4 above)
Comment (MaxWipfli) [reply Userland/Utilities/less.cpp:282]: Also, `char const*`.

File: Userland/Utilities/man.cpp (hunk 8 above)
Comment (MaxWipfli) [reply Userland/Utilities/man.cpp:147]: Couldn't `man` just `disown` the child and exit? Or does that not work because of `stdout` and friends?

Comment (alimpfard) [reply Userland/Utilities/man.cpp:147]: Hmm...
I'm not sure how that would play out with stdio's buffering and the fact that `less` doesn't read everything unless it has to...

File: Userland/Utilities/less.cpp (hunk 1 above)
Comment (Petelliott) [reply Userland/Utilities/less.cpp:157]: yep

Comment (alimpfard) [reply Userland/Utilities/less.cpp:157]: I suppose that's intentional, it'll just die with sigpipe?

Comment (Petelliott) [reply Userland/Utilities/less.cpp:157]: no, I've fixed it. it should break when read_line() is false.

File: Userland/Utilities/less.cpp (hunk 3 above)
Comment (Petelliott) [reply Userland/Utilities/less.cpp]: I now treat all unknown guards and specifiers as being false, and treat '?' and '%' at the end of the line as regular characters. I also added  '\' to escape special characters.

File: Userland/Utilities/man.cpp (hunk 8 above)
Comment (Petelliott) [reply Userland/Utilities/man.cpp:147]: I don't think the wait should be necessary, but it is for some reason. I don't know enough about job control to fix it.

File: Userland/Utilities/man.cpp (hunk 7 above)
Comment (Petelliott) [reply Userland/Utilities/man.cpp]: Although a man page with that name would have to exist for it to be exploited, I've sanitized single-quotes in the section and name.

File: Userland/Utilities/man.cpp (hunk 8 above)
Comment (Petelliott) [reply Userland/Utilities/man.cpp:152]: Is this necessary to get this merged, or is it something I can fix later?

Comment (alimpfard) [reply Userland/Utilities/man.cpp:152]: :shrug: just slap a FIXME on it and do it later :P

**Review by alimpfard — APPROVED** [review]
LGTM.

**Pull request merged**